packages/lib-webpack/src/webpack/DynamicRemotePlugin.ts

```diff
@@ -0,0 +1,167 @@
+
+export type DynamicRemotePluginOptions = Partial<{
+  /**
+   * Plugin metadata JSON file name, or the parsed plugin metadata.
+   *
+   * Default value: `plugin.json`.
+   */
+  pluginMetadata: string | PluginBuildMetadata;
+
+  /**
+   * List of extensions contributed by the plugin.
+   *
+   * The value is either a `minimatch` compatible JSON file glob pattern,
+   * or the parsed extensions array.
+   *
+   * Default value: `extensions.json`.
+   */
+  extensions: string | Extension[];
+
+  /**
+   * Modules shared between the host application and its plugins at runtime.
+   *
+   * Essential modules like `react` and `redux` will be added automatically.
+   *
+   * Default value: empty object.
+   */
+  sharedModules: WebpackSharedModules;
+}>;
+
+export class DynamicRemotePlugin implements webpack.WebpackPluginInstance {
+  private readonly pluginMetadata: PluginBuildMetadata;
+
+  private readonly extensions: Extension[];
+
+  private readonly sharedModules: WebpackSharedModules;
+
+  constructor(options: DynamicRemotePluginOptions = {}) {
+    const adaptedOptions: Required<DynamicRemotePluginOptions> = {
+      pluginMetadata: options.pluginMetadata ?? 'plugin.json',
+      extensions: options.extensions ?? 'extensions.json',
+      sharedModules: options.sharedModules ?? {},
+    };
+
+    // Load and validate plugin metadata
+    if (typeof adaptedOptions.pluginMetadata === 'string') {
+      this.pluginMetadata = loadPluginMetadata(adaptedOptions.pluginMetadata);
+    } else {
+      this.pluginMetadata = validatePluginMetadata(adaptedOptions.pluginMetadata);
+    }
+
+    // Load and validate extensions contributed by the plugin
+    if (typeof adaptedOptions.extensions === 'string') {
+      this.extensions = loadExtensions(adaptedOptions.extensions);
+    } else {
+      this.extensions = validateExtensions(adaptedOptions.extensions);
+    }
+
+    // Create plugin vs. host application shared module configuration
+    this.sharedModules = {
+      ...adaptedOptions.sharedModules,
+      ...getSharedModulesWithStrictSingletonConfig(['react', 'redux']),
+    };
+  }
+
+  apply(compiler: webpack.Compiler) {
+    const logger = compiler.getInfrastructureLogger(DynamicRemotePlugin.name);
+    const containerName = this.pluginMetadata.name;
+
+    if (!compiler.options.output.publicPath) {
+      logger.warn(
+        'output.publicPath is not defined, which may cause plugin assets to not load properly in the browser',
+      );
+    }
+
+    if (!compiler.options.output.uniqueName) {
+      logger.info(`output.uniqueName will be set to ${containerName}`);
+      compiler.options.output.uniqueName = containerName;
+    }
+
+    // Generate webpack federated module container assets
+    new webpack.container.ModuleFederationPlugin({
+      name: containerName,
+      library: {
+        type: 'jsonp',
+        name: remoteEntryCallback,
+      },
+      filename: remoteEntryScript,
```

Comment:
I would allow passing custom `filename` as well. For instance what if we want to pass `hash` or something else from the app side of things?

Reply:
As mentioned above, the `filename` should include `hash` as well.
```suggestion
      filename: options.filename ?? `${remoteEntryScript}${options.useFileHash ? `.${Date.now()}.[fullhash]` : ''}.js`,
```

Reply:
Adding hash is a good idea, will try to do this change.

Reply:
@andrewballantyne I'd suggest to do this in a follow-up to get this PR merged soon.

Basically webpack `ModuleFederationPlugin` should support placeholders like `[hash]` for the `filename` option, but I haven't tested this yet and don't want to hold merging this PR because of this.

Overall, it's a good idea to add hashes to any `.js` assets associated with the plugin. If we do this, we'll also need to pass the exact file name (with the hash) via `plugin-manifest.json` so that the host application knows its name.

Reply:
Ticket to track the work... https://issues.redhat.com/browse/HAC-771 please add any additional details or rework if I misrepresented it.